src/gui/DatabaseOpenWidget.cpp

```diff
@@ -174,28 +177,29 @@ void DatabaseOpenWidget::openDatabase()
         delete m_db;
     }
     QApplication::setOverrideCursor(QCursor(Qt::WaitCursor));
-    m_db = reader.readDatabase(&file, masterKey);
+    m_db = reader.readDatabase(&file, *masterKey);
     QApplication::restoreOverrideCursor();
 
     if (m_db) {
         if (m_ui->messageWidget->isVisible()) {
             m_ui->messageWidget->animatedHide();
         }
         emit editFinished(true);
-    }
-    else {
-        m_ui->messageWidget->showMessage(tr("Unable to open the database.")
-                                         .append("\n").append(reader.errorString()), MessageWidget::Error);
+    } else {
+        m_ui->messageWidget->showMessage(tr("Unable to open the database.").append("\n").append(reader.errorString()),
+                                         MessageWidget::Error);
         m_ui->editPassword->clear();
     }
+
+    delete masterKey;
 }
 
-CompositeKey DatabaseOpenWidget::databaseKey()
+CompositeKey* DatabaseOpenWidget::databaseKey()
 {
-    CompositeKey masterKey;
+    CompositeKey* masterKey = new CompositeKey();
```

Comment:
Shouldn't this be a smart pointer of some sort? Right now I don't see you deleting this instance anywhere.

Reply:
It could be, yeah. It's deleted line 215 if the key is invalid, otherwise it's deleted at the end of `openDatabase()`

Reply:
Im in favor of using smart/scoped pointers for these key components from here on out.

Reply:
@droidmonkey @droidmonkey now using a QScopedPointer in `openDatabase()`

Reply:
You should use and return a QSharedPointer here, not a QScopedPointer. With a QScopedPointer you have no idea when the object will be deleted and it may as well be deleted never or twice.